### DISCUSSION TITLE
How to add custom javascript in Active Admin v4 ?

### DISCUSSION THREAD

**Discussion by vccoffey** [activeadmin/activeadmin#8346] [Help]
I'm trying to add custom javascript, such as is needed for this addon
https://github.com/blocknotes/activeadmin_quill_editor

How do I do it? I've tried creating an `active_admin.js` but it is not being loaded.

**Comment by jaynetics** [comment]
AFAICT there is no built-in way at the moment, leaving only questionable options:

## monkey patching

```ruby
Importmap::ImportmapTagsHelper.prepend(Module.new do
  def javascript_importmap_tags(entry_point, ...)
    return super unless entry_point.to_s == 'active_admin'
    
    safe_join([super, YOUR_JS], "\n")
  end
end)
```

## overriding the views

- run `rails g active_admin:views`
- delete all generated views except `app/views/active_admin/_html_head.html.erb`
- in `_html_head.html.erb`, add your JS
- downside: when updating activeadmin you'll need to make sure this partial is still up to date

**Reply by javierjulio** [reply]
@jaynetics there is no need to monkey patch. Why aren't you using ActiveAdmin.importmap to pin your own JS? I don't think you understand how that works. Please keep that in mind before saying the project doesn't support adding custom JS which is false.

That's the point of the templates. It's much easier to customize and keep up to date. The contents of the html_head partial are unlikely to change. Why are you concerned about that? It gives you flexibility without depending on the library for some config. It feels more natural to just change the partial like in any Rails app.

**Reply by jaynetics** [reply]
@javierjulio can you give an example of how adding to the importmap would work? it was the first thing i tried, in quite a few ways.

**Reply by jaynetics** [reply]
> The contents of the html_head partial are unlikely to change. Why are you concerned about that?

suggesting to modify generated views might put pressure on the gem to not touch them again, because if they ever do change in a functional way, a lot of people will spend a lot of time on understanding what went wrong. having programmatic hooks in a few places would seem cleaner to me, personally, as that would allow for deprecations, enforced migration, etc.

**Reply by javierjulio** [reply]
Not at all. We've made changes to call out "template changes" in our releases for that reason. We will avoid breaking changes on the Ruby side but templates can change and will be improved. We have #8337 up next as an example of that.

The reason for templates is it makes it easier to maintain the library. We get auto loading now too when working on the library locally. Should make it easier for others to contribute and make changes in their apps. Also helps avoid monkey patching view components which was common and reduces the Arbre builder method list so you can create your own components without conflicts.

**Reply by jaynetics** [reply]
congratulations on the autoloading! i know what a big pain point that used to be.

maybe having some empty partials and rendering them in some places would suffice to give "hook" capabilities.

e.g. there could be an empty `html_head_additions` partial that is rendered inside the `html_head` partial, so users could only have that file and add their own things without overriding core things.

**Reply by tagliala** [reply]
> maybe having some empty partials and rendering them in some places would suffice to give "hook" capabilities.

I'm in favor of adding an empty partial for this use case

**Reply by javierjulio** [reply]
Sorry guys, I'm not in favor of this. Ordering can matter in the html_head for some so I prefer a simpler option of only one file that a dev can customize rather than multiple. The contents of the html_head partial is not expected to change.

**Comment by barnaclebarnes** [comment]
@javierjulio Can we add some documentation on adding custom Javascript using the active admin importmap? The documentation on the web about how to do this with a rails engine is somewhat lacking. Currently I'm doing this (which isn't the right way I'm sure).

```
# config/application.rb
Rails.application.config.importmap.paths << ActiveAdmin::Engine.root.join("config", "importmap.rb")
```

```
# app/javascripts/application.js in my main app.
import 'active_admin'
```

```
# _head.html.erb

<%= javascript_importmap_tags %>  # Replace Active Admin line.
```

What I _really_ want to do is:

1. Add a third party module to AA `pin "tom-select" # @2.3.1`
2. Create a controller in my app along the lines of https://hotwire.io/use-cases/select to provide typeahead find for a form
3. Include this controller in Active Admin

This seems like it should be doable but as I say, documentation around import camps in general is quite hard to come by. I would be happy to add a PR for documentation if I knew the "right" way to do this.

Cheers
Glen

**Reply by guuhuu** [reply]
I found @barnaclebarnes's method to be the best way right now to add custom javascript, thanks.

Appending activeadmin's importmap with my import map brakes auto-reloading of the scripts defined in the importmap
```ruby
# config/initializers/active_admin/importmap.rb

ActiveAdmin.importmap.draw(Rails.root.join("config/importmap.rb"))
```

@javierjulio Will hotwired turbo and stimulus be included in the activeadmin by default?

**Reply by javierjulio** [reply]
I went with importmap-rails after some deliberation because it had engine support and was simpler setup from the library perspective from what I recall. I understand though it's not clear on how to use it from the app side. Personally, I don't use importmap-rails. It is not required for v4 beta. You can use something else like jsbundling-rails instead which should work great. I used it initially. I think most of us on the team prefer the Node bundling approach. Feel free to use whichever is best for you. Thank you.

**Reply by thisismydesign** [reply]
@javierjulio Could you please document how to add custom JS in https://github.com/activeadmin/activeadmin/blob/master/UPGRADING.md?

`importmap-rails` is the Rails 7 way so many of us will want to use it, but it's really confusing how to make it work. In the end, I think the solution from @barnaclebarnes works the best. But I've been through hours of trying different methods that could be saved for the next person..

**Reply by agomezcampero** [reply]
Thanks for the solution @barnaclebarnes, I also had to disable turbo in ActiveAdmin to make sure everything was working correctly. In case any one else also has this issue.

```
# active_admin.html.erb

<body class="..." data-turbo="false">
```

**Reply by thisismydesign** [reply]
@javierjulio How do you use v4 without `importmap-rails`?
```
uninitialized constant #<Class:ActiveAdmin>::Importmap

[activeadmin (4.0.0.beta15) lib/active_admin.rb:66:in `importmap'](http://localhost:3000/super_admin#)
```

**Reply by nathan-appere** [reply]
@thisismydesign did you figure that one out?

**Reply by thisismydesign** [reply]
@nathan-appere No.

**Comment by guuhuu** [comment]
At this point using importmap has no benefits compared to bundling

**Reply by javierjulio** [reply]
Noted elsewhere but importmap-rails is not required. If you prefer the bundling approach (e.g. jsbundling-rails) then go for it. I think most of us on the team prefer that anyway. It should work fine. If you have issues, let us know. Thank you.

**Comment by thisismydesign** [comment]
I solved it like this: https://stackoverflow.com/questions/78985376/activeadmin-v4-how-to-add-custom-js-via-importmaps/78985377
